panoply/datasource.py

```diff
@@ -18,13 +24,58 @@ def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
```

Comment:
This is exactly the problem, it changes the source object(which is not part of this function scope). instead of returning a valid token. such functions are very difficult to debug and follow.